File: query/query/service/kerbecs.py
"""Kerbecs admin resolver client.

Maps gateway-form paths (e.g. /vehicles/123) to the concrete upstream URL by
querying the kerbecs admin /admin-gw/resolve endpoint, mirroring the Go
package gr26/pkg/kerbecs. Answers are cached for `_CACHE_TTL` seconds.
"""

from __future__ import annotations

import threading
import time

import requests
from loguru import logger

_CACHE_TTL = 300.0  # seconds

_endpoint: str | None = None
_user: str | None = None
_password: str | None = None
_cache: dict[str, tuple[str, float]] = {}
_lock = threading.RLock()


def init(endpoint: str | None, user: str | None, password: str | None) -> None:
    """Configure the resolver. Lookups happen lazily on first resolve()."""
    global _endpoint, _user, _password
    if not endpoint:
        logger.warning("KERBECS_ENDPOINT not set; service-to-service calls will fail")
        return
    _endpoint = endpoint.rstrip("/")
    _user = user
    _password = password
    logger.info(f"Kerbecs resolver configured against {_endpoint}")


def resolve(method: str, path: str) -> str:
    """Return the full upstream URL for `method path`. Raises on failure."""
    if not _endpoint:
        raise RuntimeError("kerbecs resolver not initialized")

    key = f"{method} {path}"
    now = time.monotonic()

    with _lock:
        hit = _cache.get(key)
        if hit and hit[1] > now:
            return hit[0]

    auth = (_user, _password) if _user else None
    r = requests.get(
        f"{_endpoint}/admin-gw/resolve",
        params={"path": path, "method": method},
        auth=auth,
        timeout=5,
    )
    if r.status_code == 404:
        raise RuntimeError(f"no upstream registered for {path}")
    r.raise_for_status()

    body = r.json()
    if not body.get("matched"):
        raise RuntimeError(f"no upstream registered for {path}")

    upstream_url = body["url"].rstrip("/") + body.get("rewritten_path", "")
    with _lock:
        _cache[key] = (upstream_url, now + _CACHE_TTL)
    return upstream_url

File: p987/pkg/kerbecs/kerbecs.go
// Package kerbecs resolves gateway-form paths (e.g. /vehicles/{id}) to the
// concrete upstream URL by asking the kerbecs gateway's admin /admin-gw/resolve
// endpoint. Mirrors the pattern from Sentinel — kerbecs is the routing source
// of truth, so we ask it where a request should go and cache the answer.
package kerbecs

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

const cacheTTL = 5 * time.Minute

var (
	endpoint string
	user     string
	password string
	client   = &http.Client{Timeout: 5 * time.Second}
)

type entry struct {
	url string
	exp time.Time
}

var (
	mu    sync.RWMutex
	cache = map[string]entry{}
)

// Init configures the resolver against the kerbecs admin API. No connection is
// made here — lookups happen lazily on first Resolve.
func Init(adminEndpoint, adminUser, adminPassword string) {
	endpoint = strings.TrimRight(adminEndpoint, "/")
	user = adminUser
	password = adminPassword
	go sweep()
}

type resolveResponse struct {
	Matched       bool   `json:"matched"`
	URL           string `json:"url"`
	RewrittenPath string `json:"rewritten_path"`
}

// Resolve maps a gateway-form path and HTTP method to the full upstream URL.
// Answers are cached for cacheTTL.
func Resolve(method, path string) (string, error) {
	if endpoint == "" {
		return "", fmt.Errorf("kerbecs resolver not initialized")
	}
	key := method + " " + path

	mu.RLock()
	if e, ok := cache[key]; ok && time.Now().Before(e.exp) {
		mu.RUnlock()
		return e.url, nil
	}
	mu.RUnlock()

	q := url.Values{}
	q.Set("path", path)
	q.Set("method", method)
	req, err := http.NewRequest(http.MethodGet, endpoint+"/admin-gw/resolve?"+q.Encode(), nil)
	if err != nil {
		return "", fmt.Errorf("resolve %s: %w", path, err)
	}
	req.SetBasicAuth(user, password)

	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("resolve %s: %w", path, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return "", fmt.Errorf("no upstream registered for %s", path)
	}
	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("resolve %s: kerbecs returned %d", path, resp.StatusCode)
	}

	var rr resolveResponse
	if err := json.NewDecoder(resp.Body).Decode(&rr); err != nil {
		return "", fmt.Errorf("resolve %s: decode: %w", path, err)
	}
	if !rr.Matched {
		return "", fmt.Errorf("no upstream registered for %s", path)
	}

	full := strings.TrimRight(rr.URL, "/") + rr.RewrittenPath
	mu.Lock()
	cache[key] = entry{url: full, exp: time.Now().Add(cacheTTL)}
	mu.Unlock()
	return full, nil
}

func sweep() {
	for range time.Tick(cacheTTL) {
		now := time.Now()
		mu.Lock()
		for k, e := range cache {
			if now.After(e.exp) {
				delete(cache, k)
			}
		}
		mu.Unlock()
	}
}
